### PR TITLE
minor: instance + custom emoji support with sticker picker

### PR DESCRIPTION
**Pull request by llun**
## Summary

Adds **instance custom emoji** (Mastodon `CustomEmoji`) end to end: storage, public + admin endpoints, ActivityPub federation, a postbox **sticker/emoji picker**, and an **admin management page**. Also makes `GET /api/v2/instance` public to match Mastodon.

A posted status containing `:shortcode:` now (1) renders the emoji locally, (2) federates an ActivityPub `Emoji` tag so Mastodon and other AP servers render it, and (3) is surfaced unauthenticated via `GET /api/v1/custom_emojis`.

## Endpoints — Mastodon compatibility checklist

| Endpoint | Change | Mastodon source |
|---|---|---|
| `GET /api/v1/custom_emojis` | **Now public** (removed `OAuthGuard`); returns the *listed* set: `disabled = false AND visible_in_picker = true` | `custom_emojis_controller.rb` index uses `CustomEmoji.listed`; endpoint is public ([docs](https://docs.joinmastodon.org/methods/custom_emojis/)) |
| `GET /api/v2/instance` | **Now public** (removed `OAuthGuard`) | `/methods/instance/` — served unauthenticated |
| `GET /api/v1/instance` | Unchanged (already public, shape verified) | `/methods/instance/` |
| `POST /api/v1/admin/custom_emojis` | Create via multipart upload (shared media pipeline); 422 on bad mime/dup/shortcode | mirrors Mastodon admin custom emoji create |
| `GET /api/v1/admin/custom_emojis` | List incl. disabled | admin index |
| `PATCH`/`PUT /api/v1/admin/custom_emojis/:id` | Update enable/disable, `visible_in_picker`, `category` | Mastodon admin uses `PATCH`; `PUT` added to match this repo's existing admin `[id]` convention |
| `DELETE /api/v1/admin/custom_emojis/:id` | Delete | admin destroy |

`CustomEmoji` entity fields (`shortcode`, `url`, `static_url`, `visible_in_picker`, `category`) match `custom_emoji_serializer.rb` / [entities/CustomEmoji](https://docs.joinmastodon.org/entities/CustomEmoji/). The admin shape adds `id` + `disabled`.

## ActivityPub Emoji-tag round-trip (verified on a running server)

- **Outbound:** posting `hello :blobcat:` serializes a `tag` entry `{ "type":"Emoji", "name":":blobcat:", "icon":{"type":"Image","url":…}, "id":…, "updated":… }` in the status' AP JSON (confirmed over HTTP: `GET /users/testuser/statuses/:id` with `Accept: application/activity+json`).
- **Inbound:** incoming Notes with `Emoji` tags already parse into stored `emoji` domain Tags (in `createNoteJob`); added a regression test.
- **Bug fixed:** `getMentionFromTag` previously fell through to `Mention.safeParse` for emoji tags, so the two `status.ts` `toActivityPubObject` sites emitted **bogus `Mention` objects** for emoji. It now returns `null` for emoji; new `getEmojiFromTag` emits the `Emoji` object. All three outbound builders (`getNoteFromStatus`, Note + Question in `status.ts`) include emoji uniformly.
- Mastodon REST `status.emojis` is populated automatically by the pre-existing `getEmojisFromTags`.

## How shortcodes become tags

`createNote` / `createPoll` / `updateNote` scan the text for `:shortcode:`, resolve against the **enabled** instance emoji set (ignoring `visible_in_picker`, so hand-typed non-picker emoji still federate, matching Mastodon), and persist `emoji` domain Tags. Edits re-sync (delete + recreate).

## Frontend

- **Postbox picker** (`emoji-picker-button.tsx`): one popover with **system Unicode emoji** (curated local dataset, no new deps) + **instance custom emoji** (via new `getCustomEmojis` in `lib/client.ts`), with search, category tabs, grid, and hover preview per `preview/stickers.html`. System emoji insert the character; custom emoji insert `:shortcode:` at the caret. Keyboard-accessible (labelled controls, Escape to close).
- Custom emoji render as images in the picker and in the **composer live preview** through the existing `convertEmojisToImages`/`processStatusText` pipeline (refactored to a shared `processStatusTextContent`) — no second renderer, no `dangerouslySetInnerHTML`.
- **Admin page** (`/admin/admin/emojis` under the admin section) using the design-system chrome (`SectionNavDropdown`) and the settings/stickers visual design; all network calls go through `lib/client.ts`.

## Design

Implemented from the handoff bundle (`ui_kits/web/StickerPicker.jsx`, `Stickers.jsx`, `Composer.jsx`, `SettingsPage.jsx` → `StickerSettings`): picker grid/categories/search/preview, the dashed-dropzone upload form, and list rows. Adapted to this project's Tailwind tokens and primitives; admin placement + `SectionNavDropdown` follow `CLAUDE.md`/`AGENTS.md` over the kit's settings location.

## Local verification gates

`prettier --write .` ✓ · `yarn lint` ✓ (0 errors) · `yarn build` ✓ · `yarn test` ✓ (**3545 passed / 418 suites**)

### Browser verification (local SQLite)
- Public `GET /api/v1/custom_emojis` and `GET /api/v2/instance` return 200 with **no token**.
- Picker opens, shows system categories + Custom tab; inserting a custom emoji puts `:blobcat:` in the textarea; the composer preview renders it as an inline image.
- A posted `:blobcat:` status' AP JSON contains the `Emoji` tag (checked over HTTP).
- Admin page lists/creates/updates/deletes emoji.

(Screenshots shared with the author; per `AGENTS.md` they are not committed to the repo.)

## Tests added
DB CRUD, public route (public + shape + picker filter), admin CRUD (incl. video-upload rejection), `getEmojiTags`, `getEmojiFromTag`/`getMentionFromTag`, create-note + poll shortcode→emoji-Tag resolution, outgoing Note **and** Question Emoji tag, inbound emoji-tag parsing, edit re-sync, and picker insertion (both kinds) + preview image rendering.

## Notes for reviewers
- Custom emoji uploads use the shared (actor-scoped) media pipeline via the uploading admin's session actor; an OAuth-token-only admin (no session actor) cannot upload — documented limitation.
- Animated emoji are not transcoded: `static_url = url` (uploads are PNG/JPEG only; the route rejects video/audio even though `FileSchema` allows them).
- `persistEmojiTagsForStatus` reads the enabled emoji set once per note/poll create/edit — acceptable at instance scale.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lib/types/domain/status.ts [hunk 1]
```diff
@@ -296,9 +296,10 @@ export const toActivityPubObject = (status: Status): Note | Question => {
       cc: status.cc,
       inReplyTo: status.reply || null,
       content: status.text,
-      tag: status.tags
-        .map((tag) => getMentionFromTag(tag))
-        .filter((tag) => tag !== null),
+      tag: [
+        ...status.tags.map((tag) => getMentionFromTag(tag)),
+        ...status.tags.map((tag) => getEmojiFromTag(tag))
+      ].filter((tag) => tag !== null),
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/types/domain/status.ts]: Done in f852b6b — collapsed to a single map: `getMentionFromTag(tag) ?? getEmojiFromTag(tag)` (getMentionFromTag returns null for emoji, getEmojiFromTag returns null for non-emoji, so each tag yields exactly one AP object).

File: lib/types/domain/status.ts [hunk 2]
```diff
@@ -355,9 +356,10 @@ export const toActivityPubObject = (status: Status): Note | Question => {
     attachment: originalStatus.attachments
       .filter((attachment) => !isFitnessAttachment(attachment))
       .map((attachment) => getDocumentFromAttachment(attachment)),
-    tag: originalStatus.tags
-      .map((tag) => getMentionFromTag(tag))
-      .filter((tag) => tag !== null),
+    tag: [
+      ...originalStatus.tags.map((tag) => getMentionFromTag(tag)),
+      ...originalStatus.tags.map((tag) => getEmojiFromTag(tag))
+    ].filter((tag) => tag !== null),
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/types/domain/status.ts]: Done in f852b6b — same single-map `??` rewrite applied to the Note branch here.

File: lib/utils/getNoteFromStatus.ts [hunk 3]
```diff
@@ -40,10 +40,10 @@ export const getNoteFromStatus = (
...
       .filter((attachment) => !isFitnessAttachment(attachment))
       .map((attachment) => getDocumentFromAttachment(attachment)),
-    tag: actualStatus.tags
-      .filter((tag) => tag.type !== 'emoji')
-      .map((tag) => getMentionFromTag(tag))
-      .filter((tag) => tag !== null),
+    tag: [
+      ...actualStatus.tags.map((tag) => getMentionFromTag(tag)),
+      ...actualStatus.tags.map((tag) => getEmojiFromTag(tag))
+    ].filter((tag) => tag !== null),
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/utils/getNoteFromStatus.ts]: Done in f852b6b — single-map `??` rewrite applied here too.

File: lib/components/post-box/post-box.tsx [hunk 4]
```diff
@@ -328,6 +330,52 @@ export const PostBox: FC<Props> = ({
...
+  const buildSyntheticEmojiTags = (value: string): Tag[] =>
+    getEmojiTags(value, customEmojis).map((emojiTag, index) => ({
+      id: `preview-${index}`,
+      statusId: 'preview',
+      type: 'emoji' as const,
+      name: emojiTag.name,
+      value: emojiTag.value,
+      createdAt: 0,
+      updatedAt: 0
+    }))
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/post-box/post-box.tsx]: Done in f852b6b — `buildSyntheticEmojiTags` is now wrapped in `useCallback` with `[customEmojis]`.

File: lib/components/post-box/emoji-picker-button.tsx [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+                      onClick={() => setTab(entry.id)}
+                      title={entry.name}
+                      aria-label={entry.name}
+                      aria-pressed={active}
+                      className={cn(
+                        'inline-flex size-8 shrink-0 items-center justify-center rounded-md transition-colors',
+                        active
+                          ? 'bg-primary/10 text-primary'
+                          : 'text-muted-foreground hover:bg-muted'
+                      )}
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/post-box/emoji-picker-button.tsx:204]: Done in f852b6b — added `focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring` to the category tab buttons.

File: lib/components/post-box/emoji-picker-button.tsx [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+                      onFocus={() => setHover(item)}
+                      onClick={() => pick(item)}
+                      title={item.kind === 'custom' ? item.label : item.name}
+                      aria-label={
+                        item.kind === 'custom'
+                          ? `Insert ${item.label}`
+                          : `Insert ${item.name}`
+                      }
+                      className="inline-flex aspect-square w-full items-center justify-center rounded-md transition-colors hover:bg-muted"
+                    >
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/post-box/emoji-picker-button.tsx:244]: Done in f852b6b — added the focus-visible ring classes to the emoji grid cells.

File: lib/components/admin/CustomEmojiManager.tsx [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    sortByShortcode(initialEmojis)
+  )
+  const [shortcode, setShortcode] = useState('')
+  const [category, setCategory] = useState('')
+  const [file, setFile] = useState<File | null>(null)
+  const [previewUrl, setPreviewUrl] = useState<string | null>(null)
+  const [submitting, setSubmitting] = useState(false)
+  const [error, setError] = useState<string | null>(null)
+  const [notice, setNotice] = useState<string | null>(null)
+  const fileInputRef = useRef<HTMLInputElement>(null)
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/admin/CustomEmojiManager.tsx:35]: Done in f852b6b — added a `useEffect` cleanup that revokes `previewUrl` on unmount (and `useEffect` is now imported).

File: lib/components/admin/CustomEmojiManager.tsx [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+          onSubmit={onSubmit}
+          className="grid items-start gap-4 sm:grid-cols-[auto_1fr]"
+        >
+          <div className="flex flex-col items-center gap-2">
+            <button
+              type="button"
+              onClick={() => fileInputRef.current?.click()}
+              aria-label="Choose emoji image"
+              className="flex size-24 items-center justify-center overflow-hidden rounded-xl border-2 border-dashed text-muted-foreground transition-colors hover:border-primary hover:text-primary"
+            >
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/admin/CustomEmojiManager.tsx:138]: Done in f852b6b — added the focus-visible ring to the upload dropzone button.

File: lib/components/post-box/post-box.tsx [hunk 9]
```diff
@@ -328,6 +330,52 @@ export const PostBox: FC<Props> = ({
...
+    const start = textarea.selectionStart ?? current.length
+    const end = textarea.selectionEnd ?? current.length
+    const next = current.slice(0, start) + snippet + current.slice(end)
+    onTextChange(next)
+    const caret = start + snippet.length
+    requestAnimationFrame(() => {
+      textarea.focus()
+      textarea.setSelectionRange(caret, caret)
+    })
+  }
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/post-box/post-box.tsx:380]: Not applied — wrapping insertAtCaret in useCallback would require [onTextChange] in the dep array, but onTextChange is a non-memoized const declared later in the component, so (a) evaluating it in the dep array at this call site is a temporal-dead-zone reference, and (b) it changes every render, giving the memo zero benefit. EmojiPickerButton is not memoized either, so it re-renders with the parent regardless. Left as a plain function to avoid a misleading memo. buildSyntheticEmojiTags (depends only on the stable customEmojis state) was memoized as suggested.

File: lib/components/admin/CustomEmojiManager.tsx [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+              {previewUrl ? (
+                <img
+                  src={previewUrl}
```
[the remark this answers is not in the document]
Comment (llun) [reply lib/components/admin/CustomEmojiManager.tsx]: This is a false positive for React. `previewUrl` comes from `URL.createObjectURL()` on the admin's own selected file and is rendered as a React JSX `` attribute — React escapes attribute values and never reinterprets a blob: URL as HTML (no innerHTML / dangerouslySetInnerHTML / html parser is involved). The value is a `blob:` object URL, not attacker-controlled markup, and the route is admin-only. No HTML-injection sink exists here.

**Comment by llun** [comment]
### Additional review (parallel reviewer agents) — addressed

Findings the automated review missed, now fixed:

1. **Admin create shortcode race (correctness)** — the duplicate pre-check is check-then-act; a concurrent insert would hit the unique constraint and return 500 with an orphaned media file. Fixed in `3b0931f`: the create now maps the unique-constraint error to **422** and best-effort deletes the just-saved media.
2. **Word-boundary shortcode scanning (AP/Mastodon parity)** — `getEmojiTags` used `/:([a-zA-Z0-9_]+):/g`, which would federate a shortcode embedded in a word (e.g. `foo:tada:bar`) that Mastodon's boundary-aware `CustomEmoji::SCAN_RE` would not render. Fixed in `3b0931f` with lookbehind/lookahead boundaries.
3. **Test gaps** — added in `3b0931f`: `GET /api/v2/instance` public (removed a stale passthrough `OAuthGuard` mock that masked the contract), admin `PATCH` 422 on invalid body, admin `POST` 403 when there is no session actor, edit re-sync that **adds** a new shortcode, and shortcode word-boundary cases.

Gate after all changes: prettier ✓, lint ✓ (0 errors), build ✓, test ✓ (3553 passed).

File: app/api/v1/admin/custom_emojis/schema.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  visible_in_picker: Booleanish.optional()
+})
+
+// `PATCH`/`PUT /api/v1/admin/custom_emojis/:id` — update moderation fields.
+export const UpdateCustomEmojiRequest = z.object({
+  category: z
+    .string()
+    .trim()
+    .max(255)
+    .transform((value) => value || null)
```
Comment (llun) [inline app/api/v1/admin/custom_emojis/schema.ts:45]: **Bug: clearing an emoji category returns 422 (schema rejects `null`).**

`UpdateCustomEmojiRequest.category` is `z.string().trim().max(255).transform(v => v || null).optional()`, which accepts a string or `undefined` but **not** `null`. The admin UI clears a category by sending `null` (`CustomEmojiManager` `onBlur` → `event.target.value.trim() || null`, and `adminUpdateCustomEmoji` includes the key whenever `category !== undefined`). So clearing a set category PATCHes `{"category": null}` → `safeParse` fails → 422 "Update failed", and a category can never be removed. The DB layer already persists `null` correctly; only the schema blocks it.

Reproduced: `safeParse(null)` → reject (current) vs `null` (fixed).

Fix: make it nullable while keeping `.optional()` outermost so omitted-category toggle PATCHes still preserve the existing value:
```ts
category: z.string().trim().max(255).nullable().transform((value) => value || null).optional()
```
_(severity: medium — found by review sub-agent, round 1)_

Comment (llun) [reply app/api/v1/admin/custom_emojis/schema.ts:45]: Fixed in 3388159fb — `category` is now `.nullable()` (kept `.optional()` outermost so an omitted category still preserves the existing value). Added a PATCH test asserting `{category:null}` clears it. Verified: undefined→preserve, null/''→null, 'cats'→'cats'.

File: lib/utils/text/getEmojiTags.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
+import { CustomEmojiData } from '@/lib/types/domain/customEmoji'
+
+// Matches `:shortcode:` tokens in status text. Shortcodes are limited to the
+// same character set the custom-emoji table validates (`[a-zA-Z0-9_]+`). The
+// lookbehind/lookahead require a non-alphanumeric, non-colon boundary on each
+// side, mirroring Mastodon's `CustomEmoji::SCAN_RE` so a shortcode embedded in a
+// word (e.g. `foo:bar:baz`) is not treated as an emoji and we federate exactly
+// the tokens Mastodon would render.
+export const EMOJI_SHORTCODE_REGEX =
+  /(?<![A-Za-z0-9:]):([a-zA-Z0-9_]+):(?![A-Za-z0-9:])/g
```
Comment (llun) [inline lib/utils/text/getEmojiTags.ts]: **Mastodon parity: shortcode should be `{2,}`, not `+`.**

Verified against upstream `app/models/custom_emoji.rb`: `SHORTCODE_RE_FRAGMENT = '[a-zA-Z0-9_]{2,}'`, `validates length minimum 2`, and `CustomEmoji::SCAN_RE` reuses the `{2,}` fragment. This PR allows 1-char shortcodes via `+` here (and in `CUSTOM_EMOJI_SHORTCODE_REGEX` in `lib/types/domain/customEmoji.ts`) and `.min(1)` in the admin schema. A 1-char emoji like `:a:` would be accepted, stored, and federated, but a remote Mastodon server scanner will not match it and its model rejects it — so it never renders remotely. The comment here even claims `SCAN_RE` parity, which is inaccurate for the quantifier.

Fix: use `{2,}` in `EMOJI_SHORTCODE_REGEX` and `CUSTOM_EMOJI_SHORTCODE_REGEX`, and `.min(2)` in the schema Shortcode.

_(severity: medium — review sub-agent, round 2)_

File: lib/actions/updateNote.ts [hunk 13]
```diff
@@ -58,6 +59,13 @@ export const updateNoteFromUserInput = async ({
     return null
   }
 
+  // Re-sync emoji tags when the text changes so newly added `:shortcode:`
+  // tokens federate and removed ones stop federating.
+  if (text !== undefined) {
+    await database.deleteStatusTagsByType({ statusId, type: 'emoji' })
+    await persistEmojiTagsForStatus({ database, statusId, text })
```
Comment (llun) [inline lib/actions/updateNote.ts:68]: **`updateNote` returns/timelines a status with stale emoji tags.**

`updatedStatus` is fetched at line 49 (via `database.updateNote`), *before* the emoji re-sync at lines 65–66, so the value returned to the caller and passed to `addStatusToTimelines` still carries the pre-edit emoji tags. `createNoteFromUserInput` deliberately re-fetches the status after `persistEmojiTagsForStatus`; this path does not. Federation is fine (the SEND_UPDATE job re-fetches by id), but the immediate optimistic client render and the timeline cache show a newly-added `:shortcode:` as literal text until reload.

Fix: re-fetch the status after the emoji re-sync and use that for both `addStatusToTimelines` and the return value.

_(severity: low — review sub-agent, round 2)_

File: lib/utils/text/getEmojiTags.ts (hunk 12 above)
Comment (llun) [reply lib/utils/text/getEmojiTags.ts]: Fixed in 0b93b6c80 — changed to `{2,}` in `EMOJI_SHORTCODE_REGEX` and `CUSTOM_EMOJI_SHORTCODE_REGEX`, and `.min(2)` in the admin schema, matching Mastodon's `SHORTCODE_RE_FRAGMENT`/`SCAN_RE`. Added tests rejecting a 1-char shortcode at both the scanner and admin-create levels.

File: lib/actions/updateNote.ts (hunk 13 above)
Comment (llun) [reply lib/actions/updateNote.ts:68]: Fixed in 0b93b6c80 — `updateNoteFromUserInput` now re-fetches the status after the emoji re-sync and uses it for both `addStatusToTimelines` and the return value (mirroring `createNoteFromUserInput`). Test updated to assert the returned status includes the freshly-added emoji tag.
